src/main/java/io/cdap/plugin/http/sink/batch/MessageBuffer.java

```diff
@@ -0,0 +1,232 @@
+   * @param delimiterForMessages   The delimiter to be used for messages.
+   * @param charset                The charset to be used for the message.
+   * @param customMessageBody      The custom message body to be used.
+   */
+  public MessageBuffer(
+          MessageFormatType messageFormat, String jsonBatchKey, boolean shouldWriteJsonAsArray,
+          String delimiterForMessages, String charset, String customMessageBody
+  ) {
+    this.jsonBatchKey = jsonBatchKey;
+    this.delimiterForMessages = delimiterForMessages;
+    this.charset = charset;
+    this.shouldWriteJsonAsArray = shouldWriteJsonAsArray;
+    this.customMessageBody = customMessageBody;
+    this.buffer = new ArrayList<>();
+    switch (messageFormat) {
+      case JSON:
+        messageFormatter = this::formatAsJson;
+        contentType = "application/json";
+        break;
+      case FORM:
+        messageFormatter = this::formatAsForm;
+        contentType = "application/x-www-form-urlencoded";
+        break;
+      case CUSTOM:
+        messageFormatter = this::formatAsCustom;
+        contentType = "text/plain";
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid message format: " + messageFormat);
+    }
+  }
+
+  /**
+   * Adds a record to the buffer.
+   *
+   * @param record The record to be added.
+   */
+  public void add(StructuredRecord record) {
+    buffer.add(record);
+  }
+
+  /**
+   * Clears the buffer.
+   */
+  public void clear() {
+    buffer.clear();
+  }
+
+  /**
+   * Returns the size of the buffer.
+   */
+  public int size() {
+    return buffer.size();
+  }
+
+  /**
+   * Returns whether the buffer is empty.
+   */
+  public boolean isEmpty() {
+    return buffer.isEmpty();
+  }
+
+  /**
+   * Returns the content type of the message.
+   */
+  public String getContentType() {
+    return contentType;
+  }
+
+  /**
+   * Converts the buffer to the appropriate format and returns the message.
+   */
+  public String getMessage() throws IOException {
+    return messageFormatter.apply(buffer);
+  }
+
+  private String formatAsJson(List<StructuredRecord> buffer) {
+    try {
+      return formatAsJsonInternal(buffer);
+    } catch (IOException e) {
+      throw new IllegalStateException("Error formatting JSON message. Reason: " + e.getMessage(), e);
+    }
+  }
+
+  private String formatAsJsonInternal(List<StructuredRecord> buffer) throws IOException {
+    boolean useJsonBatchKey = !Strings.isNullOrEmpty(jsonBatchKey);
+    if (!shouldWriteJsonAsArray || !useJsonBatchKey) {
+      return getBufferAsJsonList();
+    }
+    // A new StructuredRecord is created with the jsonBatchKey as the
+    // field name and the array of records as the value
+    Schema bufferRecordSchema = buffer.get(0).getSchema();
+    Schema bufferRecordArraySchema = Schema.arrayOf(bufferRecordSchema);
+    Schema wrappedMessageSchema = Schema.recordOf("wrapper",
```

Comment:
still have the comment that this schema should be created once in the constructor instead of for every method call

Reply:
Moved to constructor